### PR TITLE
[Merged by Bors] - feat(Algebra/DirectLimit): add `map, congr, lift_injective` for direct limits of modules, abelian groups and rings

### PR DESCRIPTION
**Pull request by jjaassoonn**
Consider direct limits `lim G` and `lim G'` with direct system `f` and `f'` respectively, any
family of linear maps `gᵢ : Gᵢ ⟶ G'ᵢ` such that `g ∘ f = f' ∘ g` induces a linear map
`lim G ⟶ lim G'`. When `gᵢ`'s are equivalence, so is the induced. Similarly for direct limits of groups and rings

This PR is splitted from #8473

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/DirectLimit.lean [hunk 1]
```diff
@@ -172,6 +172,99 @@ theorem lift_unique [IsDirected ι (· ≤ ·)] (F : DirectLimit G f →ₗ[R] P
...
+`lim G ⟶ lim G'`.
+-/
+def map (maps : (i : ι) → G i →ₗ[R] G' i)
+    (maps_compatible : ∀ i j h, maps j ∘ₗ f i j h = f' i j h ∘ₗ maps i) :
+    DirectLimit G f →ₗ[R] DirectLimit G' f' :=
+  lift _ _ _ _ (fun i ↦ of _ _ _ _ _ ∘ₗ maps i) fun i j h g ↦ (isEmpty_or_nonempty ι).elim
+    (fun _ ↦ Subsingleton.elim _ _) fun _ ↦ by
+      have eq1 := LinearMap.congr_fun (maps_compatible i j h) g
+      simp only [LinearMap.coe_comp, Function.comp_apply] at eq1 ⊢
+      rw [eq1, of_f]
```
Comment (eric-wieser) [inline Mathlib/Algebra/DirectLimit.lean:206]: I think we're in danger of semantically satiating `map` here; why not use the same name as you did in the docstring:
```suggestion
def map (g : (i : ι) → G i →ₗ[R] G' i) (hg : ∀ i j h, g j ∘ₗ f i j h = f' i j h ∘ₗ g i) :
    DirectLimit G f →ₗ[R] DirectLimit G' f' :=
  lift _ _ _ _ (fun i ↦ of _ _ _ _ _ ∘ₗ g i) fun i j h g ↦ (isEmpty_or_nonempty ι).elim
    (fun _ ↦ Subsingleton.elim _ _) fun _ ↦ by
      have eq1 := LinearMap.congr_fun (hg i j h) g
      simp only [LinearMap.coe_comp, Function.comp_apply] at eq1 ⊢
      rw [eq1, of_f]
```

Same throughout. In the `comp` lemmas, you can use `g\_1` and `g\_2` for the two families.

File: Mathlib/Algebra/DirectLimit.lean [hunk 2]
```diff
@@ -172,6 +172,99 @@ theorem lift_unique [IsDirected ι (· ≤ ·)] (F : DirectLimit G f →ₗ[R] P
...
+
+@[simp] lemma map_apply_of (maps : (i : ι) → G i →ₗ[R] G' i)
+    (maps_compatible : ∀ i j h, maps j ∘ₗ f i j h = f' i j h ∘ₗ maps i)
+    {i : ι} (g : G i) :
+    map maps maps_compatible (of _ _ _ _ _ g) = of R ι G' f' i (maps i g) :=
+  lift_of _ _ _
+
+lemma map_id [IsDirected ι (· ≤ ·)] :
```
Comment (eric-wieser) [inline Mathlib/Algebra/DirectLimit.lean:214]:
```suggestion
@[simp] lemma map_id [IsDirected ι (· ≤ ·)] :
```

File: Mathlib/Algebra/DirectLimit.lean [hunk 3]
```diff
@@ -350,12 +443,110 @@ theorem lift_of (i x) : lift G f P g Hg (of G f i x) = g i x :=
...
 
 theorem lift_unique [IsDirected ι (· ≤ ·)] (F : DirectLimit G f →+ P) (x) :
-    F x = lift G f P (fun i => F.comp (of G f i).toAddMonoidHom) (fun i j hij x => by simp) x := by
+    F x = lift G f P (fun i => F.comp (of G f i)) (fun i j hij x => by simp) x := by
   cases isEmpty_or_nonempty ι
   · simp_rw [Subsingleton.elim x 0, _root_.map_zero]
   · exact DirectLimit.induction_on x fun i x => by simp
 #align add_comm_group.direct_limit.lift_unique AddCommGroup.DirectLimit.lift_unique
 
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/DirectLimit.lean:452]:
```suggestion
```

**Review by eric-wieser — APPROVED** [review]
This looks great, thanks! I think I'll probably have one more round of style comments tomorrow, (or I might just make them myself)

File: Mathlib/Algebra/DirectLimit.lean (hunk 1 above)
Comment (jjaassoonn) [reply Mathlib/Algebra/DirectLimit.lean]: done

File: Mathlib/Algebra/DirectLimit.lean [hunk 4]
```diff
@@ -172,6 +172,98 @@ theorem lift_unique [IsDirected ι (· ≤ ·)] (F : DirectLimit G f →ₗ[R] P
...
+open LinearEquiv LinearMap in
+/--
+Consider direct limits `lim G` and `lim G'` with direct system `f` and `f'` respectively, any
+family of equivalences `gᵢ : Gᵢ ≅ G'ᵢ` such that `g ∘ f = f' ∘ g` induces an equivalence
+`lim G ⟶ lim G'`.
+-/
+def congr [IsDirected ι (· ≤ ·)]
+    (equivs : (i : ι) → G i ≃ₗ[R] G' i)
+    (equivs_compatible : ∀ i j h, equivs j ∘ₗ f i j h = f' i j h ∘ₗ equivs i) :
```
Comment (eric-wieser) [inline Mathlib/Algebra/DirectLimit.lean:238]:
```suggestion
family of equivalences `eᵢ : Gᵢ ≅ G'ᵢ` such that `e ∘ f = f' ∘ e` induces an equivalence
`lim G ≅ lim G'`.
-/
def congr [IsDirected ι (· ≤ ·)]
    (e : (i : ι) → G i ≃ₗ[R] G' i) (he : ∀ i j h, e j ∘ₗ f i j h = f' i j h ∘ₗ e i) :
```

Comment (jjaassoonn) [reply Mathlib/Algebra/DirectLimit.lean]: every `equiv` has been changed to `e` (I think)

File: Mathlib/Algebra/DirectLimit.lean [hunk 5]
```diff
@@ -350,12 +438,104 @@ theorem lift_of (i x) : lift G f P g Hg (of G f i x) = g i x :=
...
+    Function.Injective (lift G f P g Hg) := by
+  cases isEmpty_or_nonempty ι
+  · apply Function.injective_of_subsingleton
+  simp_rw [injective_iff_map_eq_zero] at injective ⊢
+  intros z hz
+  induction' z using DirectLimit.induction_on with _ g
+  rw [lift_of] at hz
+  rw [injective _ g hz, _root_.map_zero]
+
+section interaction_with_other_directLimits
```
Comment (eric-wieser) [inline Mathlib/Algebra/DirectLimit.lean:458]: `functorial` might be a good name for these sections

**Review by eric-wieser — APPROVED** [review]
bors d=@alreadydone

Thanks!

File: Mathlib/Algebra/DirectLimit.lean (hunk 5 above)
Comment (jjaassoonn) [reply Mathlib/Algebra/DirectLimit.lean]: thanks for the suggestion

**Review by alreadydone — APPROVED** [review]
LGTM, sorry for the delay
bors merge

I thought a bit about the possibility of using the general CategoryTheory.Limits APIs, e.g. [CategoryTheory.Limits.colim](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Limits/HasLimits.html#CategoryTheory.Limits.colim) (this is for colimit objects chosen by the axiom of choice, so we'd need to build a version with arbitrary designated colimit objects like the DirectLimits here (using [CategoryTheory.Limits.IsColimit.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Limits/IsLimit.html#CategoryTheory.Limits.IsColimit.map)). In addition, we'd need to lift the `G i` to a universe higher than the indexing type `ι`, which means we still need a version of `congr` with `e : (i : ι) → G i ≃+* G' i :=` [ULift.ringEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Ring/ULift.html#ULift.ringEquiv) etc. In the end, it won't save any work from this PR.

**Pull request closed without merging**